server/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -119,6 +123,39 @@ public NodePath(Path path) throws IOException {
             }
         }
 
+        public void addShard(ShardId shardId) {
```

Comment:
do we need to make these methods concurrency-safe?

Reply:
I have moved caller of these methods from NodeEnv to IndexService. It's synchronized now.